### PR TITLE
retro-raspberrypi: enable v4l2m2m decoder for full-kms

### PR DESCRIPTION
**Pull request by MarkusVolk**
Signed-off-by: Markus Volk <f_l_k@t-online.de>

### CODE REVIEW

File: conf/distro/include/retro-raspberrypi.inc [hunk 1]
```diff
@@ -1,16 +1,17 @@
+MACHINEOVERRIDES =. "${@bb.utils.contains('MACHINE_FEATURES', 'full-kms', 'full-kms:', '', d)}"
+
 RPI_ALWAYS_FULLHD ?= "1"
 RPI_ENABLE_AUDIO ?= "1"
+RPI_V4L2M2M_DECODER_rpi_full-kms ?= "1"
```
Comment (dev-0x7C6) [inline conf/distro/include/retro-raspberrypi.inc:5]: Using `?=` operator with conditional assignment is undefined from what I know (only `=` operator is allowed). 
I think that `rpi` override can be drop for this variable in favor of simple `?=` assignment.

like: 
`RPI_V4L2M2M_DECODER ?= "${@bb.utils.contains('OVERRIDES', 'full-kms', '1', '0', d)}`

**Comment by dev-0x7C6** [comment]
Is this already working with kodi/ffmpeg, looks exiting :)

File: conf/distro/include/retro-raspberrypi.inc (hunk 1 above)
Comment (dev-0x7C6) [reply conf/distro/include/retro-raspberrypi.inc:5]: > Using ?= operator with conditional assignment is undefined from what I know (only = operator is allowed).

It happens to work correctly, I tested something like that 

recipe:
```bash
FOO_rpi ??= "1"
FOO_rpi ?= "2" 
FOO_rpi = "3"
# order test
FOO_rpi ?= "4"
FOO_rpi ??= "5"
```
output:
```
~/.../master/retro-build-environment/build ⑂master* $ bitbake -e conditional-assignment-experiment | grep ^FOO
FOO="3"
FOO_rpi="3"
```

Comment (dev-0x7C6) [reply conf/distro/include/retro-raspberrypi.inc:5]: After couple of thoughts, beside testing assignment condition.

`RPI_V4L2M2M_DECODER ?=` should be preferred:
* `RPI_V4L2M2M_DECODER = "0"` in `local.conf` will be overridden by assignment condition even this condition is using `?=` or `??=`
* It's hard to change default value with current approach user need to track our overrides and put something like that `RPI_V4L2M2M_DECODER_rpi_full-kms = "0"` in `local.conf`
* `RPI_V4L2M2M_DECODER_rpi_full-kms = "0"` also creates a problem that we need to be backward compatible with exact overrides.

**Comment by MarkusVolk** [comment]
I have used the override because I could only test it on the Raspberry pi 4. But you're probably right, it shouldn't matter if RPI_V4L2M2M_DECODER = "1" was set, even for older models.

I only tested very briefly, but it did work with patches in Kodi. I was able to play 1080p videos with good frame rates and a cpu load of 20%. However, it wasn't completely stable. Kodi occasionally died when using the drmprime decoder. In Kodi, the decoder must be activated under Settings -> Player -> Video (you need at least level 'Advanced' to display the option).

The patches for video acceleration are taken from Libreelec. Amazing work is done there

I have a work-in-progress layer that I use to build my image for the Rock Pi. There still has to be a lot of fixing and cleanup, but as it is, at least a 'retro-image standard' build for Rockchip and Raspberry should be possible

The following layers need to be added to bblayers.conf:

  $ {YOCTOROOT} / layers / meta-browser / meta-chromium \
  $ {YOCTOROOT} / layers / meta-openembedded / meta-filesystems \
  $ {YOCTOROOT} / layers / meta-openembedded / meta-gnome \
  $ {YOCTOROOT} / layers / meta-openembedded / meta-xfce \
  $ {YOCTOROOT} / layers / meta-retro-wayland \
  $ {YOCTOROOT} / layers / meta-rockchip-extras \


Kodi, Retroarch and a current Chromium are being built for wayland / sway. Video acceleration also works for the Rock Pi, although that's not really necessary. The RK3399 can play 1080p content in Kodi and Chromium even without a HW decoder.

I'll push my work-in-progress meta-retro-wayland layer, so you could have a look at and do some testing if you like. I also created a meta-rockchip-extras layer to include some kernel patches and needed firmware for wifi / bt. Maybe you want to include it to the project?

**Comment by MarkusVolk** [comment]
https://github.com/MarkusVolk/meta-retro-wayland.git
https://github.com/MarkusVolk/meta-rockchip-extras.git

i use this settings in custom.conf:
```

DISTRO_FEATURES_append += " \
        gobject-introspection-data \
        gtk+3 \
        polkit \
        bluetooth \
        opengl \
        x11 \
        wayland \
        alsa \
        pulseaudio \
        retroarch-online \
        retroarch-updater \
        rauc \
        retroarch-automount \
        retroarch-firmware \
        retroarch-gles3 \ 
"

IMAGE_FEATURES_append += "ssh-server-openssh"
IMAGE_FEATURES_remove += "ssh-server-dropbear"

IMAGE_INSTALL_remove = "retroarch-service"
PREFERRED_VERSION_waylandpp = "0.2.8"
```

**Comment by dev-0x7C6** [comment]
> I'll push my work-in-progress meta-retro-wayland layer, so you could have a look at and do some testing if you like. I also created a meta-rockchip-extras layer to include some kernel patches and needed firmware for wifi / bt.

It's looking impressive, I'll build and research your work. 

> Maybe you want to include it to the project?

You want expand meta-retro with additional stuff or you mean extending build environment with extra layers and configuration? Anyway my initial goal with meta-retro was create libretro cores collection with retroarch frontend, but I endup expanding with distro configuration add additions. I don't mind to make it full blown media center distro, I'm open for changes.

I even started my own application loader project [here](https://github.com/dev-0x7C6/retro-menu), but there is no progress for some time. Maybe there is already better alternative.

As always, thanks for patches :)

**Comment by MarkusVolk** [comment]
Thought about adding meta-rockchip-extras to retro-build-environment. It adds support for 4k resolution, makes wifi/bt and rkvdec working. Things that might be useful even without a full blown media distribution.
For the stuff inside meta-retro-wayland ... this indeed goes way beyond the scope of a reference distribution. Thats why i created it as an additional layer. But if you like the idea of full media support I'm also open for merging this into meta-retro

**Comment by dev-0x7C6** [comment]
> Thought about adding meta-rockchip-extras to retro-build-environment. It adds support for 4k resolution, makes wifi/bt and rkvdec work.

Added.

> For the stuff inside meta-retro-wayland ... this indeed goes way beyond the scope of a reference distribution. Thats why i created it as an additional layer. But if you like the idea of full media support I'm also open for merging this into meta-retro

Let's proceed as is, I'll keep eye on meta-retro-wayland to get more idea what can be shared and how makes things easier to integrate with.

**Pull request merged**